src/boot/efi/devicetree.c

```diff
@@ -104,6 +113,90 @@ EFI_STATUS devicetree_install(struct devicetree_state *state, EFI_FILE *root_dir
                         MAKE_GUID_PTR(EFI_DTB_TABLE), PHYSICAL_ADDRESS_TO_POINTER(state->addr));
 }
 
+static const char* devicetree_get_compatible(const void *dtb) {
+        assert(dtb);
+        const struct fdt_header *dt_header = dtb;
+        if (!IS_ALIGNED64(dt_header) || be32toh(dt_header->Magic) != UINT32_C(0xd00dfeed))
+                return NULL;
+
+        uint32_t dt_size = be32toh(dt_header->TotalSize);
+        uint32_t struct_off = be32toh(dt_header->OffDTStruct);
+        uint32_t struct_size = be32toh(dt_header->SizeDTStruct);
+        uint32_t strings_off = be32toh(dt_header->OffDTStrings);
+        uint32_t strings_size = be32toh(dt_header->SizeDTStrings);
+        uint32_t tmp;
+
+        if (struct_off % sizeof(uint32_t) != 0 || struct_size % sizeof(uint32_t) != 0 ||
+            check_add_overflow(strings_off, strings_size, &tmp) || tmp > dt_size ||
+            check_add_overflow(struct_off, struct_size, &tmp) || tmp > strings_off)
+                return NULL;
+
+        const uint32_t *cursor = (const uint32_t *) ((uint8_t *) dt_header + struct_off);
+        const char *strings_block = (const char *) ((uint8_t *) dt_header + strings_off);
+
+        size_t size_words = struct_size / sizeof(uint32_t);
+        if (check_sub_overflow(size_words, 3, &tmp))
+                return NULL;
+
+        for (size_t i = 0; i < tmp; i++) {
+                switch (be32toh(cursor[i])) {
+                case FDT_BEGIN_NODE:
+                        if (cursor[++i] != 0)
+                                return NULL;
+                        break;
+                case FDT_NOP:
+                        break;
+                case FDT_PROP:
+                        size_t len = be32toh(cursor[++i]);
+                        size_t name_off = be32toh(cursor[++i]);
+                        size_t len_words = DIV_ROUND_UP(len, sizeof(uint32_t));
+                        size_t s;
+
+                        if (!check_add_overflow(name_off, strlen8("compatible"), &s) &&
+                            s < strings_size && streq8(strings_block + name_off, "compatible")) {
+                                const char *c = (const char *) &cursor[++i];
+                                if (len == 0 || i + len_words > size_words || c[len - 1] != '\0')
+                                        c = NULL;
+
+                                return c;
+                        }
+                        i += len_words;
+                        break;
+                default:
+                        return NULL;
+
+                }
+        }
+
+        return NULL;
+}
+
+EFI_STATUS devicetree_match(const void *dtb_buffer, size_t dtb_length) {
+        assert(dtb_buffer);
+        const struct fdt_header *dt_header = (const struct fdt_header *)dtb_buffer;
+
```

Comment:
Please add a comment here that describes what this function does in a high-level language, i.e. what is being matched with what, and maybe even how those strings could look. This will make it easier for future readers.